src/diff/index.js

```diff
@@ -24,13 +24,6 @@ import options from '../options';
  * Fragments that have siblings. In most cases, it starts out as `oldChildren[0]._dom`.
  */
 export function diff(parentDom, newVNode, oldVNode, context, isSvg, excessDomChildren, mounts, ancestorComponent, force, oldDom) {
-	// If the previous type doesn't match the new type we drop the whole subtree
```

Comment:
These checks are not necessary when `diff` is always called by `diffChildren` (there is one exception, `forceUpdate`, discussed below).  `diffChildren` already does proper null checking on newVNode and passes in EMPTY_OBJ if oldVNode is `null`. `diffChildren` will never pass two vnodes of non-equal types. It will instead always pass `null` if it can't find a matching oldVNode for newVNode.

The only place `diff` is called not inside `diffChildren` is `forceUpdate`. However, `forceUpdate` always calls `diff` with non-null vnodes of the same type. It will never pass in a null newVNode or oldVNode cuz it always passes the VNode that represents the component. If the component is in the tree, it can't have a null VNode or else it doesn't exist in the tree. It can render null, but that is handled by `diffChildren`.

In other words, all of this logic is duplicated in `diffChildren`. Since we now call always `diffChildren` after `diff`, we no longer need to duplicate that logic here.